src/cogent3/core/new_genetic_code.py

```diff
@@ -0,0 +1,438 @@
+
+
+def _make_converter(
+    kmer_alpha: new_alphabet.KmerAlphabet, codons: tuple[str, ...], code_sequence: str
+) -> typing.Callable[[bytes, bytes], bytes]:
+    assert len(codons) == len(code_sequence) == 64
+    # we get the index of the codon in the kmer alphabet
+    kmers = numpy.array(
+        [kmer_alpha.to_index(codon) for codon in codons], dtype=numpy.uint8
+    )
+    return new_alphabet.convert_alphabet(kmers.tobytes(), code_sequence.encode("utf8"))
+
+
+@dataclasses.dataclass
+class GeneticCode:
+    """Holds codon to amino acid mapping, and vice versa."""
+
+    ID: int
+    name: str
+    ncbi_code_sequence: dataclasses.InitVar[str]
+    ncbi_start_codon_map: dataclasses.InitVar[str]
+    moltype: new_moltype.MolType = new_moltype.DNA
+    _codon_to_aa: typing.Dict[str, str] = dataclasses.field(init=False, default=None)
+    _aa_to_codon: typing.Dict[str, typing.List[str]] = dataclasses.field(
+        init=False, default=None
+    )
+    _sense_codons: SetStr = dataclasses.field(init=False, default=None)
+    _stop_codons: SetStr = dataclasses.field(init=False, default=None)
+    _start_codons: SetStr = dataclasses.field(init=False, default=None)
+    codons: new_alphabet.KmerAlphabet = dataclasses.field(init=False, default=None)
+    anticodons: typing.Tuple[str, ...] = dataclasses.field(init=False, default=None)
+    # callables for translating on the plus strand, or the minus strand
+    _translate_plus: ConverterType = dataclasses.field(init=False, default=None)
+    _translate_minus: ConverterType = dataclasses.field(init=False, default=None)
+
+    def __post_init__(self, ncbi_code_sequence: str, ncbi_start_codon_map: str):
+        trinuc_alpha = self.moltype.alphabet.get_kmer_alphabet(k=3, include_gap=False)
+        self._codon_to_aa, self._aa_to_codon, self._stop_codons = _make_mappings(
+            trinuc_alpha, ncbi_code_sequence
+        )
+        self.codons = trinuc_alpha
+        self._start_codons = {
+            self.codons[i] for i in _get_start_codon_indices(ncbi_start_codon_map)
+        }
+        self._sense_codons = tuple(
+            c for c in self.codons if self._codon_to_aa[c] != "*"
+        )
+        self.anticodons = tuple(self.moltype.rc(codon) for codon in self.codons)
+        self._translate_plus = _make_converter(
+            trinuc_alpha, self.codons, ncbi_code_sequence
+        )
+        self._translate_minus = _make_converter(
+            trinuc_alpha, self.anticodons, ncbi_code_sequence
+        )
+
+    @property
+    def stop_codons(self) -> SetStr:
+        return self._stop_codons
+
+    @property
+    def start_codons(self) -> SetStr:
+        return self._start_codons
+
+    @property
+    def sense_codons(self) -> SetStr:
+        return self._sense_codons
+
+    def to_table(self):
+        """returns aa to codon mapping as a cogent3 Table"""
+        from cogent3.core.moltype import IUPAC_PROTEIN_code_aa
```

Comment:
any reason why import isn't placed at the top of the file? Also refer to `moltype.IUPAC_PROTEIN_code_aa` in-line?

Reply:
should be importing from new_moltype?

Reply:
Doh!!